### DISCUSSION TITLE
Can I use Oceananigans.jl finite-volume functionality for a different package?

### DISCUSSION THREAD

**Discussion by navidcy** [CliMA/Oceananigans.jl#1766] [Computational science!]
I was wondering whether it would be feasible to use Oceananigans.jl finite-volume functionality to solve different equations. The equations I have in mind are not necessarily ocean/climate related so I don't think it's appropriate to add a model in Oceananigans.jl. What I'm thinking is somehow re-exporting some module of Oceananigans.jl in a different package and taking it off from there?

Thoughts?

**Comment by glwagner** [comment]
Yes, in short. 

https://github.com/ali-ramadhan/Atmosfoolery.jl does just that.

The "basics":

https://github.com/CliMA/Oceananigans.jl/blob/383173d11a0c96182a4349fc1e33755207bf0886/src/Oceananigans.jl#L166-L177

are really tools (grids, functions / operators, fields, and solvers) for finite volume calculations.

It could be fun to build a model / time-stepping framework using abstract operations + `ComputedField` (rather than hand-writing as we do in Oceananigans). I don't think we know if this would be performant or not --- but it might.

**Comment by francispoulin** [comment]
Thanks @glwagner for sharing this.  

I had a similar idea for doing some MHD.  I don't think that a model would fit at all in Oceananigans but I think it's very easy to do with Oceananigams.  For example, for the `ShallowWaterModel`, you could add a tracer, say `A`, to be the magnetic streamfunction and then add the Lorentz force that depends on `A` and the velocity.  It's an example that I plan to put together but I haven't yet.

@navidcy : are you thinking about solving different model equations?

**Reply by navidcy** [reply]
Yeap, MHD was one idea. E.g., setups like those in doi:[10.1103/PhysRevFluids.4.083701](https://doi.org/10.1103/PhysRevFluids.4.083701) but not necessarily periodic...

Other idea was some mixed layer-thermocline conceptual model we are putting together with @dhruvbhagtani2105 (but it's on the works still).

**Reply by francispoulin** [reply]
Sounds like fun and happy to chat sometime if you want, since we both seem to want to use Oceananigans to study MHD.
